sklearn/tree/export.py

```diff
@@ -397,6 +398,23 @@ def recurse(tree, node_id, criterion, parent=None, depth=0):
             return_string = True
             out_file = six.StringIO()
 
+        # Check length of feature_names before get into the tree node
+        # raise error for too few feature_names
+        if feature_names is not None:
+            if len(feature_names) < decision_tree.n_features_:
+                raise ValueError("Length of feature_names=%d "
+                                 "does not match number of features=%d"
+                                 % (len(feature_names),
+                                    decision_tree.n_features_))
+            # for too much feature_names, will use the first n_features
+            # raise an warning for users
+            if len(feature_names) > decision_tree.n_features_:
```

Comment:
Good point. My vote would be still to be raise an error and document it as a bug-fix over here (https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new.rst)

Something on the lines of `export_graphviz` now raises an error if length of feature_names is not the same as the number of features.